core/src/main/java/com/linecorp/armeria/common/TlsProvider.java

```diff
@@ -0,0 +1,127 @@
+
+import io.netty.handler.ssl.SslContextBuilder;
+
+/**
+ * Provides {@link TlsKeyPair}s for TLS handshakes.
+ */
+@UnstableApi
+@FunctionalInterface
+public interface TlsProvider {
+
+    /**
+     * Returns a {@link TlsProvider} which always returns the specified {@link TlsKeyPair}.
+     */
+    static TlsProvider of(TlsKeyPair tlsKeyPair) {
+        requireNonNull(tlsKeyPair, "tlsKeyPair");
+        return new StaticTlsProvider(tlsKeyPair);
+    }
+
+    /**
+     * Returns a newly created {@link ServerTlsProviderBuilder}.
+     *
+     * <p>Example usage:
+     * <pre>{@code
+     * TlsProvider
+     *   .builderForServer()
+     *   // Set the default key pair.
+     *   .setDefault(TlsKeyPair.of(...))
+     *   // Set the key pair for "example.com".
+     *   .set("example.com", TlsKeyPair.of(...))
+     *   .build();
+     * }</pre>
+     */
+    static ServerTlsProviderBuilder builderForServer() {
+        return new ServerTlsProviderBuilder();
+    }
+
+    /**
+     * Returns a newly created {@link ClientTlsProviderBuilder}.
+     * <pre>{@code
+     * TlsProvider
+     *   .builderForClient()
+     *   // Set the key pair for "api.example.com".
+     *   .set("api.example.com", TlsKeyPair.of(...))
+     *   // Set the key pair for "web.example.com".
+     *   .set("web.example.com", TlsKeyPair.of(...))
+     *   .build();
+     * }</pre>
+     */
+    static ClientTlsProviderBuilder builderForClient() {
+        return new ClientTlsProviderBuilder();
+    }
+
+    /**
+     * Finds a {@link TlsKeyPair} for the specified {@code hostname}.
+     *
+     * <p>If no matching {@link TlsKeyPair} is found for a hostname, {@code "*"} will be specified to get the
+     * default {@link TlsKeyPair}.
+     * If no default {@link TlsKeyPair} is found, {@code null} will be returned.
+     */
+    @Nullable
+    TlsKeyPair find(String hostname);
+
+    /**
+     * Allows the bad cipher suites listed in
+     * <a href="https://datatracker.ietf.org/doc/html/rfc7540#appendix-A">RFC7540</a> for TLS handshake.
+     *
+     * <p>Note that enabling this option increases the security risk of your connection.
+     * Use it only when you must communicate with a legacy system that does not support
+     * secure cipher suites.
+     * See <a href="https://datatracker.ietf.org/doc/html/rfc7540#section-9.2.2">Section 9.2.2, RFC7540</a> for
+     * more information. This option is disabled by default.
+     *
+     * @deprecated It's not recommended to enable this option. Use it only when you have no other way to
+     *             communicate with an insecure peer than this.
+     */
+    @Deprecated
+    default boolean allowsUnsafeCiphers() {
+        return false;
+    }
+
+    /**
+     * Returns the {@link Consumer} which can arbitrarily configure the {@link SslContextBuilder} that will be
+     * applied to the SSL session.
+     */
+    default Consumer<SslContextBuilder> tlsCustomizer() {
+        return builder -> {};
+    }
```

Comment:
That will move the responsibility of managing the life cycle of `SslContext` to `TlsProvider`, which may complicate things.

Reply:
There are several difficulties in doing so.

- `TlsProvider` is designed to be used at both client and server. However, `SslContext` has different properties for them, e.g., `SslContextBuilder.forClient()` and `SslContextBuilder.forServer()`. `SslContext` for a client should not be returned for a server.
- Creating a `SslContext` in Armeria optimized for HTTP/2 and TLS 1.3. Users who implement other own `TlsProvder` may feel difficult to build `SslContext` or we need to expose a factory method to easily create it. Instead, creating `TlsKeyPair` is straightforward.

Reply:
> That will move the responsibility of managing the life cycle of SslContext to TlsProvider, which may complicate things.

What do you think of modifying the method for creating the SSL context so that the `TlsProvider` doesn't have to manage the lifecycle at all? Here's a proposed interface change:
```java
interface TlsProvider {
  SslContext createSslContext(String hostname, SessionProtocol sessionProtocol, TlsEngineType tlsEngineType);
}
```
I'm suggesting this because it seems like the current `TlsProvider` functions more as a data class than as a provider.

> TlsProvider is designed to be used at both client and server.

At present, we can use `TlsProvider.ofServer()` for the client side as well. By validating the `TlsProvider` when it's assigned to a client or a server, we can prevent this misuse.
If you still prefer to use `TlsProvider` for both, we could add enum parameters to `TlsEngineType` such as `CLIENT_JDK`, `SERVER_JDK`, `CLIENT_OPENSSL`, etc.

> Users who implement their own TlsProvider may find it difficult to build SslContext, or we might need to expose a factory method to make it easier.

That is true. I also worry about exposing Netty's SslContext API to the public. However, creating an SslContext might not be too difficult since Netty provides a well-designed SslContextBuilder API. To address the issue of exposing the API, we could provide our own abstraction, similar to what we did with `DnsEndpointGroupBuilder`.

Reply:
I'm still questioning about exposing an API that creates `SslContext`.
- The API forces all users who want to implement `TlsProvider` to learn how to build `SslContext`.
- With the current API, users only need to convert their certificates using `TlsKeyPair`'s factory method for basic usage.
- Advanced users may change SslContext using the customizer.

I prefer the current style because would be more convenient for most users.

Reply:
> That will move the responsibility of managing the life cycle of SslContext to TlsProvider, which may complicate things.

I'm not sure what the lifecycle of `SslContext` is - do you mean it is expensive to create an `SslContext` and you prefer that users don't have to worry about caching logic?

I think there are a couple of points here:

##### On the argument of returning `SslContext` vs `TlsKeyPair`

Honestly I'm find with either way. Later on, I think we can allow a higher level API if users want

##### On `TlsProvider` methods

Say that I am a user implementing a custom `TlsProvider` - I'm not sure if it is possible to gracefully change certificates.

e.g. If I want from `TlsKeyPair1` where `allowsUnsafeCiphers=true` should be guaranteed, to `TlsKeyPair2` where `allowsUnsafeCiphers=false`, can I do this atomically? Is it possible that `TlsKeyPair2` is sent with `allowsUnsafeCiphers=true`?

```
public interface TlsProvider {
  boolean allowsUnsafeCiphers()
  TlsKeyPair find(String hostname)
}
```

Reply:
Good point. I split the TLS configuration parts from `TlsProvider` and added `ServerTlsConfig` and `ClientTlsConfig`. Those configurations are immutable objects which return a constant `allowsUnsafeCiphers` and other values.